hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/RatisUnhealthyReplicationCheckHandler.java

```diff
@@ -0,0 +1,193 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hdds.scm.container.replication.health;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.hdds.scm.container.ContainerInfo;
+import org.apache.hadoop.hdds.scm.container.ReplicationManagerReport;
+import org.apache.hadoop.hdds.scm.container.replication.ContainerCheckRequest;
+import org.apache.hadoop.hdds.scm.container.replication.ContainerHealthResult;
+import org.apache.hadoop.hdds.scm.container.replication.RatisContainerReplicaCount;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+import static org.apache.hadoop.hdds.protocol.proto.HddsProtos.ReplicationType.RATIS;
+
+/**
+ * This class handles UNHEALTHY RATIS replicas. Its responsibilities include:
+ * <ul>
+ *   <li>If only unhealthy replicas exist, check if there are replication
+ *   factor number of replicas and queue them for under/over replication if
+ *   needed.
+ *   </li>
+ *   <li>If there is perfect replication of healthy replicas, remove any
+ *   excess unhealthy replicas.
+ *   </li>
+ * </ul>
+ */
+public class RatisUnhealthyReplicationCheckHandler extends AbstractCheck {
+  public static final Logger LOG = LoggerFactory.getLogger(
+      RatisUnhealthyReplicationCheckHandler.class);
+
+  @Override
+  public boolean handle(ContainerCheckRequest request) {
+    if (request.getContainerInfo().getReplicationType() != RATIS) {
+      // This handler is only for Ratis containers.
+      return false;
+    }
+    ReplicationManagerReport report = request.getReport();
+    ContainerInfo container = request.getContainerInfo();
+
+    /*
+    First, verify there's perfect replication without considering UNHEALTHY
+    replicas.
+     */
+    if (!verifyPerfectReplication(request)) {
+      return false;
+    }
+
+    // Now, consider UNHEALTHY replicas when calculating replication status
+    RatisContainerReplicaCount replicaCount =
+        new RatisContainerReplicaCount(container,
+            request.getContainerReplicas(), request.getPendingOps(),
+            request.getMaintenanceRedundancy(), true);
+    if (replicaCount.getUnhealthyReplicaCount() == 0) {
+      LOG.info("No UNHEALTHY replicas are present for container {} with " +
+          "replicas [{}].", container, request.getContainerReplicas());
```

Comment:
This should probably be a debug log. Looks like it would get printed for every perfectly replicated container without unhealthy replicas.

Reply:
Yea I agree this should be debug.

Reply:
Done